apps/src/templates/AppView.jsx

```diff
@@ -0,0 +1,23 @@
+'use strict';
+
+var ProtectedStatefulDiv = require('./ProtectedStatefulDiv.jsx');
+
+/**
+ * Top-level React wrapper for our standard blockly apps.
+ */
+var AppView = React.createClass({
+  propTypes: {
+    renderCodeApp: React.PropTypes.func.isRequired,
+    onMount: React.PropTypes.func.isRequired
+  },
+
+  componentDidMount: function () {
+    this.refs.codeApp.getDOMNode().innerHTML = this.props.renderCodeApp();
+    this.props.onMount();
+  },
+
+  render: function () {
+    return <ProtectedStatefulDiv ref="codeApp" />;
+  }
```

Comment:
Oh yeah, not sure why I didn't just pass the render method as a property to the `ProtectedStatefulDiv`.  That's a good idea.

I'm not sure I 100% understand the scenario and additional validation you're describing, so let me echo it to see if we're on the same page.

You're saying the `ProtectedStatefulDiv` guards against unmounting but not against forgetting to mount it in the first place?  That's true, but in practice I don't know how problematic it is.

Let's consider a your code mode/design mode example.  Both modes contain a `ProtectedStatefulDiv` at some level (code mode because of droplet, design mode because of the stateful work area.

To write your top-level view to safely allow switching between the two, it must always render both (using style to hide one or the other).

```jsx
render: function () {
  var show = {display: 'block'};
  var hide = {display: 'none'};
  var codeStyle = this.props.mode === 'CODE' ? show : hide;
  var designStyle = this.props.mode === 'DESIGN' ? show : hide;
  return <div>
    <CodeMode style={codeStyle} />
    <DesignMode style={designStyle} />
  </div>;
}
```

Using this model, you'll never fail to render either, even on initial render, regardless of whether you load in code mode or design mode.

I imagine you're picturing a scenario more like this:

```jsx
render: function () {
  if (this.props.mode === 'CODE') {
    return <div><CodeMode /></div>;
  } else if (this.props.mode === 'DESIGN') {
    return <div><DesignMode /></div>;
  }
  return <div></div>;
}
```

But that model, while it would succeed on initial render (ignoring the fact that studioApp.init would blow up because it depends on all sorts of stuff existing after that initial render is done) will throw as soon as you switch modes because it would unmount one component or the other.  So I'm having trouble thinking of a practical scenario in which this is likely to go unnoticed.

Are you wanting the top-level view to somehow also assert that all of its descendant `ProtectedStatefulDiv`s have been rendered when it's initially mounted?  You could give it special knowledge about all the stateful areas in its descendant components so it can do that validation.  The downside to this is that the dev has to remember to add that validation any time they introduce a new stateful area, so you're equally vulnerable to developer error as we are now - I'm not sure it adds much protection over what we've got now.